### ISSUE TITLE
Create custom widget

### ISSUE THREAD

**Issue by Sonnenspeer** [Mapsui/Mapsui#436]
Hi there!

How is it possible to add an own custom widget to the map and render it?

I followed the sample of the zoom widget and created a custom widget that inherits from Widget class. It seems like there is a need of a custom widget renderer but as far as I have understood 
the custom renderer will never be called.

Related to: https://github.com/Mapsui/Mapsui/blob/c60b7551bc0b6b4ade24669eda2c352237b7e171/Mapsui.Rendering.Xaml/WidgetRenderer.cs#L26

Is that right? Are custom widgets not possible at the moment?

**Comment by charlenni** [comment]
You have to create WidgetRenderers for the different renderers (Xaml, Skia), that implement the static function Draw and than add a line to each WidgetRenderer, so that this draw function is called.

In the moment this only works by changing the code. @pauldendulk works on a interface, so that it is possible to create widgets without changing the code.

What type of widget do you want to create?

**Comment by pauldendulk** [comment]
I made a start with a widget renderer plugins in skia (not yet wpf). It was the first thing that came to mind. The interface will probably change. You can register a renderer for a widget by adding it to the MapControl.WidgetRenderer list.

Rendering now goes like this:
https://github.com/Mapsui/Mapsui/blob/master/Mapsui.Rendering.Skia/WidgetRenderer.cs#L18

Registering goes like this:
https://github.com/Mapsui/Mapsui/blob/master/Mapsui.Rendering.Skia/MapRenderer.cs#L40

The skia renderer instance is not accessible right now. In the WPF MapControl you could case the Renderer to the skia MapRenderer to register you own controls.

**Comment by Sonnenspeer** [comment]
I wanted to create a button that checks the geolocation of the current device (gps) and navigates the map to this point.

I'm working in Xamarin Forms so WPF is not an option for me.

**Comment by charlenni** [comment]
You don't need a Widget to do this. Create a Button and place this over the MapView with a AbsolutLayout or a Grid. Than set the event of this Button to a function, which change the center of the map with NavigateTo() to your center.

If you want a MyLocationLayer, which could display a image for your location than you could find an example at the [Xamarin.Forms MapControl](https://github.com/charlenni/Mapsui/blob/Forms/Mapsui.UI.Forms/Objects/MyLocationLayer.cs). If I look into the code, than I think it needs some update, because Mapsui has now some of the functions implemented (SVG images) ;)

**Comment by charlenni** [comment]
@pauldendulk Looks very nice and easy to use. But shouldn't there be an insanity check like
````
renders[widget.GetType()]?.Draw(canvas, screenWidth, screenHeight, widget, layerOpacity);
````

**Comment by pauldendulk** [comment]
Definately.

**Comment by charlenni** [comment]
You have even to do
````
if (widget != null)
 {
    renders[widget.GetType()]?.Draw(canvas, screenWidth, screenHeight, widget, layerOpacity);
}
````
because `widgets` could hold null widgets. They are created, when a layer hasn't an attribution.

**Comment by charlenni** [comment]
Btw. perhaps you shouldn't create a widget for layers with empty attribution.

**Comment by pauldendulk** [comment]
I added this check yesterday? https://github.com/Mapsui/Mapsui/commit/d641492a5d55dedf238bce972c0dfae7ca285c12

Is this what you mean?

**Comment by charlenni** [comment]
Yes, that is what I mean. You updated this 12 hours ago and I updated my local repository in the mornig when I start to work. This was 14 hours ago ;)
